tests/NTest/NTestEnv.lua

```diff
@@ -0,0 +1,60 @@
+-- A series of convenient utility functions for use with NTest on NodeMCU
+-- devices under test (DUTs).
+
+local NTE = {}
+
+function NTE.hasC(m)
+  -- this isn't great, but it's what we've got
+  local mstr = node.info('build_config').modules
+  return mstr:match("^"..m..",") -- first module in the list
+      or mstr:match(","..m.."$") -- last module in the list
+      or mstr:match(","..m..",") -- somewhere else
+      or error("Missing C module " .. m)
+end
+
+function NTE.hasL(m)
+  return pcall(require, m) or error("Missing Lua module " .. m)
+end
+
+function NTE.hasEnv(reqFeats)
+  -- Stream through our configuration file and extract the features attested
+  --
+  -- We expect the configuration file to attest features as keys in a dictionary
+  -- so that they can be efficiently probed here but also so that we can
+  -- parameterize features.
+  --
+  -- { "features" : {
+  --    "feat1" : ... ,
+  --    "feat2" : ...
+  --   }
+  -- }
+  local decoder = sjson.decoder({
+    metatable = {
+      checkpath = function(t, p)
+        if #p > 2 then return true end              -- something we're keeping
+        if #p == 0 then return true end             -- root table
+        if p[1] == "features" then
+          if #p == 1 then return true end           -- features table
+          local thisFeat = p[2]
+          assert type(thisFeat) == "string"
+          for k, v in ipairs(reqFeats) do
+            if v == thisFeat then return true end   -- requested feature
+          end
+        end
+        return false
+      end
+    }
+  })
+  local cfgf = file.open("testenv.conf", "r")
+  local cstr
+  repeat cstr = cfgf:read(); decoder:write(cstr) until #cstr == 0
+  cfgf:close()
+  local givenFeats = decoder:result().features
```

Comment:
Maybe check for no features table at all

Reply:
My initial hunch is to check by `assert`ing, but that does raise the question of ergonomics of `hasEnv` for prefligh tests... we'd have to `pcall` them to route exceptions to `skip`.  Can we change `NTest` just a little in some way to indicate that some `.test()` functions should treat exceptions as `Bail out!` and not `not ok`?  I'm not sure what's maximally ergonomic here and so am quite open to suggestions.

Reply:
If occurs to me that it's most likely that these `NTestEnv` functions are likely to be used before the first `test` call in a NTest file.  By way of example, the ADC-via-the-I2C-expander test will probably look like

```
local N = -[[ ... ]]
local E = require "NTestEnv"

assert hasC("adc")
assert hasC("i2c")
local config = assert(E.getConfig("adc", "i2c"))

N.test(-[[ ... ]])
-- ...
```

Reply:
given that hasEnv (or (hasFeat) now returns a boolean it is easier now.
`if not hasEnv("myrequirement") the skip("some optional message") end`

But what I was requesting is to handle the case that there is no `features` section in the config file at all. Meaning to treat it as if it was empty.

Reply:
Does the new `assert` for the `features` table not do the right thing?

Reply:
> If occurs to me that it's most likely that these `NTestEnv` functions are likely to be used before the first `test` call in a NTest file. By way of example, the ADC-via-the-I2C-expander test will probably look like
> 
> ```
> local N = -[[ ... ]]
> local E = require "NTestEnv"
> 
> assert E.hasC("adc")
> assert E.hasC("i2c")
> assert E.hasL("mcp23017")
> local config = assert(E.getConfig("adc", "mcp23017", "i2c"))
> 
> N.test(-[[ ... ]])
> -- ...
> ```

That would abort th run outside of NTest which would make it harder to use TAP.
I understood so far that these functions would be called in the first test which works as preflight.
Something like this:

```
local N = -[[ ... ]]
local E = require "NTestEnv"

local config

N.test("preflight", function ...
  if not E.hasConfig("adc", i2c", "mcp23017") then bailOut() end
  local config = E.getConfig("adc", "mcp23017", "i2c")
end)
```
Not really sure about the details yet, so if you have a better Idea ...

Reply:
> Does the new `assert` for the `features` table not do the right thing?

Yes of coarse. Now I see it. It does even better in failing early with a clear error message.